spring-integration-core/src/main/java/org/springframework/integration/util/FunctionIterator.java

```diff
@@ -16,30 +16,68 @@
 
 package org.springframework.integration.util;
 
-import java.io.Closeable;
-import java.io.IOException;
 import java.util.Iterator;
 import java.util.function.Function;
 
+import org.springframework.lang.Nullable;
+
 /**
  * An {@link Iterator} implementation to convert each item from the target
  * {@link #iterator} to a new object applying the {@link #function} on {@link #next()}.
  *
  * @author Artem Bilan
  * @author Ruslan Stelmachenko
+ * @author Gary Russell
  * @since 4.1
  */
 public class FunctionIterator<T, V> implements CloseableIterator<V> {
 
+	private final AutoCloseable closeable;
+
 	private final Iterator<T> iterator;
 
 	private final Function<? super T, ? extends V> function;
 
+	/**
+	 * Construct an instance with the provided iterable and function.
+	 * @param iterable the iterable.
+	 * @param function the function.
+	 */
 	public FunctionIterator(Iterable<T> iterable, Function<? super T, ? extends V> function) {
-		this(iterable.iterator(), function);
+		this(null, iterable.iterator(), function);
+	}
+
+	/**
+	 * Construct an instance with the provided root object, iterable and function.
+	 * @param closeable an {@link AutoCloseable} to close when iteration is complete.
+	 * @param iterable the iterable.
+	 * @param function the function.
+	 * @since 5.0.7
+	 */
+	public FunctionIterator(@Nullable AutoCloseable closeable, Iterable<T> iterable,
```

Comment:
This `closeable` is not propagated to the target ctor.
Will fix on merge.

Reply:
Doh - thanks.